doc/website/blog/2022-03-31-ent-hooks-examples.md

```diff
@@ -0,0 +1,411 @@
+as close as possible to the entity. Therefore hooks are just that - a piece of code that runs before and after an entity 
+is changing.
+
+## Let's do some examples
+
+For our examples we use a simple schema of Users and Dogs, each user can have multiple pets and each pet has an owner 
+as can be seen here:
+```go title="ent/schema/user.go"
+// User holds the schema definition for the User entity.
+type User struct {
+	ent.Schema
+}
+
+// Fields of the User.
+func (User) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("name").
+			NotEmpty(),
+		field.String("phone_number").
+			NotEmpty(),
+		field.String("last_digits").
+			Sensitive().
+			NotEmpty(),
+	}
+}
+
+// Edges of the User.
+func (User) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("pets", Dog.Type),
+		edge.To("cache", Cache.Type).Unique(),
+	}
+}
+```
+```go title="ent/schema/dog.go"
+// Dog holds the schema definition for the Dog entity.
+type Dog struct {
+	ent.Schema
+}
+
+// Fields of the Dog.
+func (Dog) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("name").
+			NotEmpty(),
+		field.Int("owner_id").
+			Optional(),
+	}
+}
+
+// Edges of the Dog.
+func (Dog) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("owner", User.Type).
+			Ref("pets").
+			Field("owner_id"). // Field is needed later.
+			Unique(),
+	}
+}
+```
+
+For each example I'll provide a test and a recipe, I highly suggest starting with a test to see that the hook does what you need it to do.
+Hooks should do one thing only and therfore be simple to test.
+
+### Keeping our secrets safe:
+
+Many times, we need to store strings that include secrets, a good example can be a phone number.
+In our case the privacy concern is rather simple. We require to hide the last 4 digits of each phone number.
+The expected behavior is best described by this test:
+```go title="ent/schema/user_test.go"
+func TestUserPhoneNumberHook(t *testing.T) {
+	ctx := context.Background()
+	c := enttest.Open(t, dialect.SQLite, "file:TestSchemaConfHooks?mode=memory&cache=shared&_fk=1")
+	u := c.User.Create().SetName("Yoni").SetPhoneNumber("315-194-6020").SaveX(ctx)
+	require.Equal(t, "315-194-****", u.PhoneNumber)
+	require.Equal(t, "6020", u.LastDigits)
+	require.Equal(t, "315-194-6020", u.FullPhoneNumber())
+}
+```
+
+As you can see, inserting a phone number updates the string to have `****` instead of the last digits and stores them 
+in the `last_digits` field that is defined as [sensitive](https://entgo.io/docs/schema-fields#sensitive-fields).
+To have the full phone number we created a utility function `FullPhoneNumber()`.
+
+#### Recipe:
+
+First thing, add the hook to your schema:
+```go title="ent/schema/user.go"
+// Hooks of the User.
+func (User) Hooks() []ent.Hook {
+	return []ent.Hook{
+		hook.If(maskPhoneNumber,
+			hook.And(hook.Or(hook.HasOp(ent.OpUpdateOne), hook.HasOp(ent.OpCreate)), hook.HasFields(user.FieldPhoneNumber)),
+		),
```

Comment:
```suggestion
		hook.If(maskPhoneNumber,
			hook.And(hook.Or(hook.HasOp(ent.OpUpdateOne), hook.HasOp(ent.OpCreate)), hook.HasFields(user.FieldPhoneNumber)),
		),
```

format it, and maybe explain about the hooks helpers that are used here. 

Also, use `hook.HasOp(ent.OpCreate | ent.OpUpdateOne)` instead of `hook.Or`.